### PR TITLE
MCOL-641 Fixes for arithmetic operations.

### PR DESCRIPTION
**Pull request by tntnatbry**
1. Perform type promotion to wide decimal if the result
   of an arithmetic operation has a precision > 18.
2. Only set the decimal width of an arithmetic operation to wide
   if both the LHS and RHS of the operation are decimal types.

### CODE REVIEW

File: dbcon/mysql/ha_mcs_execplan.cpp [hunk 1]
```diff
@@ -3601,68 +3601,55 @@ ArithmeticColumn* buildArithmeticColumn(
 
     // @bug5715. Use InfiniDB adjusted coltype for result type.
     // decimal arithmetic operation gives double result when the session variable is set.
-    CalpontSystemCatalog::ColType mysql_type = colType_MysqlToIDB(item);
+    CalpontSystemCatalog::ColType mysqlType = colType_MysqlToIDB(item);
 
-    if (mysql_type.colDataType == CalpontSystemCatalog::DECIMAL ||
-        mysql_type.colDataType == CalpontSystemCatalog::UDECIMAL)
+    CalpontSystemCatalog::ColType leftColType = pt->left()->data()->resultType();
```
Comment (drrtuy) [inline dbcon/mysql/ha_mcs_execplan.cpp]: resultType() returns a const IIRC so leftColType can be a const symbol.

File: dbcon/mysql/ha_mcs_execplan.cpp [hunk 2]
```diff
@@ -3601,68 +3601,55 @@ ArithmeticColumn* buildArithmeticColumn(
...
+    CalpontSystemCatalog::ColType rightColType = pt->right()->data()->resultType();
+
+    // Only tinker with the type if all columns involved are decimal
+    if ((mysqlType.colDataType == CalpontSystemCatalog::DECIMAL ||
```
Comment (drrtuy) [inline dbcon/mysql/ha_mcs_execplan.cpp:3610]: Plz make an ORed comparison an inline function to reduce a boilerplate code.

**Review by drrtuy — COMMENTED** [review]
Could you ellaborate on the goals behind the change in tupleaggregatestep.cpp ?

File: dbcon/mysql/ha_mcs_execplan.cpp (hunk 1 above)
Comment (tntnatbry) [reply dbcon/mysql/ha_mcs_execplan.cpp]: Done.

Comment (tntnatbry) [reply dbcon/mysql/ha_mcs_execplan.cpp]: Also replaced the OR code with an inline isDecimalOperands().

**Comment by tntnatbry** [comment]
There are 2 changes to tupleaggregastep.cpp:
  1. Addition of a for loop condition to avoid a potential segfault in the case `scale.size() < jobInfo.nonConstCols.size()` or `scale.size()  < jobInfo.nonConstDelCols.size()`
  2. Replaced `oidsAgg.push_back()`, `keysAgg.push_back()` calls in `wideDecimalOrLongDouble()` which were duplicated in both the if and else blocks, with a single set of calls in the calling function.

**Comment by drrtuy** [comment]
> 2. Replaced `oidsAgg.push_back()`, `keysAgg.push_back()` calls in `wideDecimalOrLongDouble()` which were duplicated in both the if and else blocks, with a single set of calls in the calling function.

The second change divides assignment into parts so one can easily overlook this. I would suggest to put keysAgg.push_back() and oidsAgg.push_back() into the function after if-else.

File: dbcon/mysql/ha_mcs_execplan.cpp (hunk 1 above)
Comment (drrtuy) [reply dbcon/mysql/ha_mcs_execplan.cpp]: Thx a lot.

**Review by drrtuy — APPROVED** [review]

**Pull request merged**